File: src/zellij_tab_namer/naming.py
"""Pure naming policy for Zellij pane state."""

from __future__ import annotations

from dataclasses import dataclass
import os
import re
import shlex
from typing import Any, Callable, Dict, Iterable, List, Mapping, Optional, Tuple


Compressor = Callable[[str, int], Optional[str]]

_DEFAULT_TAB_RE = re.compile(r"^(?:Tab #\d+|\d+)$")
_SPACE_RE = re.compile(r"\s+")


@dataclass(frozen=True)
class NamerConfig:
    max_chars: int = 32
    fallback_separator: str = " - "


@dataclass(frozen=True)
class RenameDecision:
    tab_id: int
    current_name: str
    candidate: str
    reason: str
    skipped: bool = False
    skip_reason: Optional[str] = None


def plan_renames(
    panes: Iterable[Mapping[str, Any]],
    state: Optional[Mapping[str, Any]] = None,
    config: Optional[NamerConfig] = None,
    compressor: Optional[Compressor] = None,
    force: bool = False,
) -> Tuple[List[RenameDecision], Dict[str, Dict[str, str]]]:
    """Return rename decisions and the generated-name state after applying them."""

    config = config or NamerConfig()
    generated = dict((state or {}).get("generated", {}))
    next_state = {"generated": generated}
    grouped = _group_panes_by_tab(panes)
    decisions: List[RenameDecision] = []

    for tab_id in sorted(grouped):
        tab_panes = grouped[tab_id]
        chosen = choose_pane(tab_panes)
        if chosen is None:
            continue

        current_name = _string(chosen.get("tab_name"))
        candidate, reason = label_for_pane(chosen, config, compressor=compressor)
        if not candidate:
            decisions.append(
                RenameDecision(
                    tab_id=tab_id,
                    current_name=current_name,
                    candidate="",
                    reason="none",
                    skipped=True,
                    skip_reason="no_candidate",
                )
            )
            continue

        tab_key = str(tab_id)
        last_generated = generated.get(tab_key)
        manual_override = (
            not force
            and not is_default_tab_name(current_name)
            and current_name != last_generated
            and current_name != candidate
        )

        if manual_override:
            decisions.append(
                RenameDecision(
                    tab_id=tab_id,
                    current_name=current_name,
                    candidate=candidate,
                    reason=reason,
                    skipped=True,
                    skip_reason="manual_override",
                )
            )
            continue

        if current_name == candidate:
            generated[tab_key] = candidate
            decisions.append(
                RenameDecision(
                    tab_id=tab_id,
                    current_name=current_name,
                    candidate=candidate,
                    reason=reason,
                    skipped=True,
                    skip_reason="already_named",
                )
            )
            continue

        generated[tab_key] = candidate
        decisions.append(
            RenameDecision(
                tab_id=tab_id,
                current_name=current_name,
                candidate=candidate,
                reason=reason,
            )
        )

    return decisions, next_state


def choose_pane(panes: Iterable[Mapping[str, Any]]) -> Optional[Mapping[str, Any]]:
    valid = [pane for pane in panes if _is_eligible_pane(pane)]
    if not valid:
        return None
    return max(valid, key=_pane_score)


def label_for_pane(
    pane: Mapping[str, Any],
    config: NamerConfig,
    compressor: Optional[Compressor] = None,
) -> Tuple[str, str]:
    title = clean_title(_string(pane.get("title")))
    command = command_name(_string(pane.get("pane_command")))

    if is_useful_title(title, command):
        return _compress_or_shorten(title, config.max_chars, compressor), "title"

    fallback = config.fallback_separator.join(
        part for part in (project_name(_string(pane.get("pane_cwd"))), command) if part
    )
    if fallback:
        return shorten_label(fallback, config.max_chars), "fallback"
    return "", "none"


def clean_title(title: str) -> str:
    collapsed = _SPACE_RE.sub(" ", title).strip()
    while collapsed and not collapsed[0].isalnum() and collapsed[0] not in {"~", "/", "."}:
        collapsed = collapsed[1:].lstrip()
    return collapsed


def shorten_label(label: str, max_chars: int) -> str:
    normalized = _SPACE_RE.sub(" ", label).strip()
    if max_chars < 1:
        return ""
    if len(normalized) <= max_chars:
        return normalized
    if max_chars <= 3:
        return normalized[:max_chars]

    limit = max_chars - 3
    prefix = normalized[:limit].rstrip()
    word_cut = prefix.rfind(" ")
    if word_cut >= max(8, limit // 2):
        prefix = prefix[:word_cut].rstrip()
    if not prefix:
        prefix = normalized[:limit]
    return f"{prefix}..."


def is_default_tab_name(name: str) -> bool:
    stripped = name.strip()
    return not stripped or bool(_DEFAULT_TAB_RE.match(stripped))


def is_useful_title(title: str, command: str = "") -> bool:
    if not title:
        return False

    lower_title = title.lower()
    lower_command = command.lower()
    generic_titles = {
        "bash",
        "zsh",
        "fish",
        "sh",
        "vim",
        "nvim",
        "hx",
        "helix",
        "claude code",
        "codex",
    }
    if lower_title in generic_titles:
        return False
    if lower_command and lower_title == lower_command:
        return False
    if _looks_like_path(title):
        return False
    return True


def command_name(command: str) -> str:
    if not command:
        return ""
    try:
        parts = shlex.split(command)
    except ValueError:
        parts = command.split()
    if not parts:
        return ""
    return os.path.basename(parts[0]) or parts[0]


def project_name(cwd: str) -> str:
    if not cwd:
        return ""
    stripped = cwd.rstrip("/")
    if not stripped:
        return ""
    return os.path.basename(stripped)


def _group_panes_by_tab(panes: Iterable[Mapping[str, Any]]) -> Dict[int, List[Mapping[str, Any]]]:
    grouped: Dict[int, List[Mapping[str, Any]]] = {}
    for pane in panes:
        tab_id = pane.get("tab_id")
        if tab_id is None:
            continue
        try:
            tab_id_int = int(tab_id)
        except (TypeError, ValueError):
            continue
        grouped.setdefault(tab_id_int, []).append(pane)
    return grouped


def _is_eligible_pane(pane: Mapping[str, Any]) -> bool:
    return not bool(pane.get("is_plugin")) and not bool(pane.get("exited"))


def _pane_score(pane: Mapping[str, Any]) -> int:
    title = clean_title(_string(pane.get("title")))
    command = command_name(_string(pane.get("pane_command")))
    if is_useful_title(title, command):
        score = 100
    elif title and _looks_like_path(title):
        score = 20
    else:
        score = 10

    if bool(pane.get("is_focused")):
        score += 3
    if not bool(pane.get("is_floating")):
        score += 2

    return score


def _compress_or_shorten(
    label: str, max_chars: int, compressor: Optional[Compressor]
) -> str:
    if len(label) <= max_chars:
        return label
    if compressor is not None:
        compressed = compressor(label, max_chars)
        if compressed:
            cleaned = clean_title(compressed)
            if cleaned and len(cleaned) <= max_chars:
                return cleaned
    return shorten_label(label, max_chars)


def _looks_like_path(title: str) -> bool:
    return (
        title.startswith("~/")
        or title.startswith("/")
        or "/." in title
        or ("/" in title and " " not in title)
    )


def _string(value: Any) -> str:
    if value is None:
        return ""
    return str(value)

File: rust/src/naming.rs
use std::collections::HashMap;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PaneSnapshot {
    pub title: String,
    pub command: String,
    pub cwd: String,
    pub focused: bool,
    pub floating: bool,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum RenameDecision {
    Rename(String),
    AlreadyNamed,
    SkipManual,
    NoCandidate,
}

#[derive(Default)]
pub struct NamingState {
    generated: HashMap<usize, String>,
}

impl NamingState {
    pub fn decision(
        &mut self,
        tab_id: usize,
        current_name: &str,
        pane: &PaneSnapshot,
        max_chars: usize,
    ) -> RenameDecision {
        let candidate = label_for_pane(pane, max_chars);
        if candidate.is_empty() {
            return RenameDecision::NoCandidate;
        }

        let last_generated = self.generated.get(&tab_id);
        let is_manual = !is_default_tab_name(current_name)
            && last_generated.map(String::as_str) != Some(current_name)
            && current_name != candidate;
        if is_manual {
            return RenameDecision::SkipManual;
        }

        self.generated.insert(tab_id, candidate.clone());
        if current_name == candidate {
            RenameDecision::AlreadyNamed
        } else {
            RenameDecision::Rename(candidate)
        }
    }

    pub fn retain_tabs(&mut self, active_tab_ids: &[usize]) {
        self.generated
            .retain(|tab_id, _| active_tab_ids.contains(tab_id));
    }
}

pub fn choose_pane(panes: &[PaneSnapshot]) -> Option<&PaneSnapshot> {
    panes.iter().max_by_key(|pane| pane_score(pane))
}

pub fn label_for_pane(pane: &PaneSnapshot, max_chars: usize) -> String {
    let title = clean_title(&pane.title);
    let command = command_name(&pane.command);
    if is_useful_title(&title, &command) {
        return shorten_label(&title, max_chars);
    }

    let project = project_name(&pane.cwd).or_else(|| project_name(&title));
    let fallback = match (project, command.as_str()) {
        (Some(project), "") => project.to_owned(),
        (Some(project), command) => format!("{project} - {command}"),
        (None, command) => command.to_owned(),
    };
    shorten_label(&fallback, max_chars)
}

pub fn is_default_tab_name(name: &str) -> bool {
    let name = name.trim();
    if name.is_empty() || name.chars().all(|character| character.is_ascii_digit()) {
        return true;
    }
    name.strip_prefix("Tab #")
        .is_some_and(|suffix| !suffix.is_empty() && suffix.chars().all(|c| c.is_ascii_digit()))
}

fn clean_title(title: &str) -> String {
    let collapsed = title.split_whitespace().collect::<Vec<_>>().join(" ");
    collapsed
        .trim_start_matches(|character: char| {
            !character.is_alphanumeric() && !matches!(character, '~' | '/' | '.')
        })
        .trim()
        .to_owned()
}

fn is_useful_title(title: &str, command: &str) -> bool {
    if title.is_empty() || looks_like_path(title) {
        return false;
    }
    let lower = title.to_ascii_lowercase();
    let is_default_pane_name = lower
        .strip_prefix("pane #")
        .is_some_and(|suffix| !suffix.is_empty() && suffix.chars().all(|c| c.is_ascii_digit()));
    !matches!(
        lower.as_str(),
        "bash" | "zsh" | "fish" | "sh" | "vim" | "nvim" | "hx" | "helix" | "claude code" | "codex"
    ) && !is_default_pane_name
        && lower != command.to_ascii_lowercase()
}

fn pane_score(pane: &PaneSnapshot) -> usize {
    let title = clean_title(&pane.title);
    let command = command_name(&pane.command);
    let mut score = if is_useful_title(&title, &command) {
        100
    } else if looks_like_path(&title) {
        20
    } else {
        10
    };
    if pane.focused {
        score += 3;
    }
    if !pane.floating {
        score += 2;
    }
    score
}

fn command_name(command: &str) -> String {
    let executable = command.split_whitespace().next().unwrap_or_default();
    executable
        .rsplit(['/', '\\'])
        .next()
        .unwrap_or_default()
        .to_owned()
}

fn project_name(path: &str) -> Option<&str> {
    if !looks_like_path(path) {
        return None;
    }
    path.trim_end_matches('/')
        .rsplit('/')
        .next()
        .filter(|name| !name.is_empty())
}

fn looks_like_path(value: &str) -> bool {
    value.starts_with("~/")
        || value.starts_with('/')
        || value.contains("/.")
        || (value.contains('/') && !value.contains(' '))
}

fn shorten_label(label: &str, max_chars: usize) -> String {
    let normalized = label.split_whitespace().collect::<Vec<_>>().join(" ");
    let length = normalized.chars().count();
    if max_chars == 0 {
        return String::new();
    }
    if length <= max_chars {
        return normalized;
    }
    if max_chars <= 3 {
        return normalized.chars().take(max_chars).collect();
    }

    let limit = max_chars - 3;
    let prefix: String = normalized.chars().take(limit).collect();
    let word_cut = prefix.rfind(' ');
    let prefix = match word_cut {
        Some(index) if index >= 8.max(limit / 2) => prefix[..index].trim_end(),
        _ => prefix.trim_end(),
    };
    format!("{prefix}...")
}

#[cfg(test)]
mod tests {
    use super::*;

    fn pane(title: &str, command: &str, cwd: &str) -> PaneSnapshot {
        PaneSnapshot {
            title: title.into(),
            command: command.into(),
            cwd: cwd.into(),
            focused: true,
            floating: false,
        }
    }

    #[test]
    fn useful_title_wins_over_shell_fallback() {
        let input = pane(
            "Implement native WASM tab naming",
            "/opt/homebrew/bin/fish",
            "/Users/eduardoleal/eduardoleal/zellij-tab-namer",
        );

        assert_eq!(label_for_pane(&input, 28), "Implement native WASM...");
    }

    #[test]
    fn generic_title_falls_back_to_project_and_activity() {
        let input = pane(
            "fish",
            "/opt/homebrew/bin/fish",
            "/Users/eduardoleal/.config/zellij",
        );

        assert_eq!(label_for_pane(&input, 32), "zellij - fish");
    }

    #[test]
    fn default_pane_title_falls_back_to_project_and_activity() {
        let input = pane(
            "Pane #1",
            "/opt/homebrew/bin/fish",
            "/Users/eduardoleal/.config/zellij",
        );

        assert_eq!(label_for_pane(&input, 32), "zellij - fish");
    }

    #[test]
    fn useful_non_floating_pane_beats_focused_generic_floating_pane() {
        let mut floating = pane("fish", "fish", "/tmp");
        floating.floating = true;
        let mut task = pane("Fix permission cache", "codex", "/tmp");
        task.focused = false;

        assert_eq!(
            choose_pane(&[floating, task]).unwrap().title,
            "Fix permission cache"
        );
    }

    #[test]
    fn manual_name_is_preserved_until_cleared() {
        let mut state = NamingState::default();
        let input = pane("New task title", "codex", "/tmp/project");

        assert_eq!(
            state.decision(4, "prod deploy", &input, 32),
            RenameDecision::SkipManual
        );
        assert_eq!(
            state.decision(4, "Tab #5", &input, 32),
            RenameDecision::Rename("New task title".into())
        );
    }

    #[test]
    fn generated_names_update_without_churn() {
        let mut state = NamingState::default();
        let first = pane("First task", "codex", "/tmp/project");
        let second = pane("Second task", "codex", "/tmp/project");

        assert_eq!(
            state.decision(5, "Tab #6", &first, 32),
            RenameDecision::Rename("First task".into())
        );
        assert_eq!(
            state.decision(5, "First task", &first, 32),
            RenameDecision::AlreadyNamed
        );
        assert_eq!(
            state.decision(5, "First task", &second, 32),
            RenameDecision::Rename("Second task".into())
        );
    }
}
